File: kotlin-node/karakum/src/jsMain/resources/plugins/convertIterableIterator.js
import ts from "typescript";

export default function (node) {
    if (
        ts.isIdentifier(node)
        && node.text === "IterableIterator"

        && node.parent
        && ts.isTypeReferenceNode(node.parent)
        && node.parent.typeName === node
    ) {

        return "js.iterable.JsIterator"
    }
    if (
        ts.isIdentifier(node)
        && node.text === "AsyncIterableIterator"

        && node.parent
        && ts.isTypeReferenceNode(node.parent)
        && node.parent.typeName === node
    ) {

        return "js.iterable.AsyncIterator"
    }
    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertIterableIterator.kt
package node.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.util.getParentOrNull
import typescript.isIdentifier
import typescript.isTypeReferenceNode

val convertIterableIterator = createPlugin { node, _, _ ->
    nullable {
        ensure(isIdentifier(node))
        ensure(node.text == "IterableIterator")

        val typeReference = ensureNotNull(node.getParentOrNull())
        ensure(isTypeReferenceNode(typeReference))
        ensure(typeReference.typeName === node)

        "js.iterable.JsIterator"
    } ?: nullable {
        ensure(isIdentifier(node))
        ensure(node.text == "AsyncIterableIterator")

        val typeReference = ensureNotNull(node.getParentOrNull())
        ensure(isTypeReferenceNode(typeReference))
        ensure(typeReference.typeName === node)

        "js.iterable.AsyncIterator"
    }
}
